<launch>
  <!-- Load parameters from the YAML file -->
  <rosparam command="load" file="$(find multi_robot_system)/config/data_extraction.yaml"/>
  <arg name="enableRviz" default="true"/>
  <arg name="mrs" default="true"/>
  <arg name="cola2_architecture" default="true"/>
  <!-- <arg name="number_of_auvs" default="6"/> -->

   <!-- Robot name arguments -->
  <arg name="robot_0" default="robot0"/>
  <arg name="robot_1" default="robot1"/>      
  <arg name="robot_2" default="robot2"/>  
  <arg name="robot_3" default="robot3"/>  
  <arg name="robot_4" default="robot4"/>
  <arg name="robot_5" default="robot5"/>
  <arg name="robot_6" default="robot6"/>

  <!-- COLA2 architecture -->
  <group if="$(arg cola2_architecture)">
    <!-- Static transform -->
    <node name="world2ned" pkg="tf" type="static_transform_publisher" args="0 0 0 0 0 3.1415 world world_ned 100"/>

    <!-- Include AUVs -->
    <include file="$(find cola2_turbot)/launch/sim_start.launch" if="$(eval int(arg('number_of_auvs')) >= 1)">
        <arg name="robot_name" value="$(arg robot_0)"/>
        <arg name="enable_keyboard" value="true"/>
    </include>
    <include file="$(find cola2_turbot)/launch/sim_start.launch" if="$(eval int(arg('number_of_auvs')) >= 2)">
        <arg name="robot_name" value="$(arg robot_1)"/>
        <arg name="enable_keyboard" value="true"/>
    </include>
    <include file="$(find cola2_turbot)/launch/sim_start.launch" if="$(eval int(arg('number_of_auvs')) >= 3)">
        <arg name="robot_name" value="$(arg robot_2)"/>
        <arg name="enable_keyboard" value="true"/>
    </include>
    <include file="$(find cola2_turbot)/launch/sim_start.launch" if="$(eval int(arg('number_of_auvs')) >= 4)">
        <arg name="robot_name" value="$(arg robot_3)"/>
        <arg name="enable_keyboard" value="true"/>
    </include>
    <include file="$(find cola2_turbot)/launch/sim_start.launch" if="$(eval int(arg('number_of_auvs')) >= 5)">
        <arg name="robot_name" value="$(arg robot_4)"/>
        <arg name="enable_keyboard" value="true"/>
    </include>
    <include file="$(find cola2_turbot)/launch/sim_start.launch" if="$(eval int(arg('number_of_auvs')) >= 6)">
        <arg name="robot_name" value="$(arg robot_5)"/>
        <arg name="enable_keyboard" value="true"/>
    </include>

    <!-- Include ASV -->
    <include file="$(find cola2_xiroi)/launch/sim_start.launch" if="$(eval int(arg('number_of_auvs')) == 2)">
        <arg name="robot_name" value="$(arg robot_2)"/>
        <arg name="safety" value="false"/>
        <arg name="enable_gui" value="false"/>
        <arg name="enable_keyboard" value="true"/>
    </include>
    <include file="$(find cola2_xiroi)/launch/sim_start.launch" if="$(eval int(arg('number_of_auvs')) == 3)">
        <arg name="robot_name" value="$(arg robot_3)"/>
        <arg name="safety" value="false"/>
        <arg name="enable_gui" value="false"/>
        <arg name="enable_keyboard" value="true"/>
    </include>
    <include file="$(find cola2_xiroi)/launch/sim_start.launch" if="$(eval int(arg('number_of_auvs')) == 4)">
        <arg name="robot_name" value="$(arg robot_4)"/>
        <arg name="safety" value="false"/>
        <arg name="enable_gui" value="false"/>
        <arg name="enable_keyboard" value="true"/>
    </include>
    <include file="$(find cola2_xiroi)/launch/sim_start.launch" if="$(eval int(arg('number_of_auvs')) == 5)">
        <arg name="robot_name" value="$(arg robot_5)"/>
        <arg name="safety" value="false"/>
        <arg name="enable_gui" value="false"/>
        <arg name="enable_keyboard" value="true"/>
    </include>
    <include file="$(find cola2_xiroi)/launch/sim_start.launch" if="$(eval int(arg('number_of_auvs')) == 6)">
        <arg name="robot_name" value="$(arg robot_6)"/>
        <arg name="safety" value="false"/>
        <arg name="enable_gui" value="false"/>
        <arg name="enable_keyboard" value="true"/>
    </include>
  </group>

  <!-- Multi-robot system -->
  <group if="$(arg mrs)">
    <group ns="/mrs">
      <!-- Load configuration files -->
      <rosparam command="load" file="$(find multi_robot_system)/config/multi_robot_system.yaml"/>
      <rosparam command="load" file="$(find multi_robot_system)/config/data_extraction.yaml"/>

      <!-- Dynamically add AUV nodes -->
      <node name="AUV_$(arg robot_0)" pkg="multi_robot_system" type="AUV_robot.py" output="screen" respawn="true" if="$(eval int(arg('number_of_auvs')) >= 1)">
        <param name="robot_ID" value="0"/>
        <param name="robot_name" value="$(arg robot_0)"/>
        <param name="navigation_depth" value="5"/>
      </node>
      <node name="AUV_$(arg robot_1)" pkg="multi_robot_system" type="AUV_robot.py" output="screen" respawn="true" if="$(eval int(arg('number_of_auvs')) >= 2)">
        <param name="robot_ID" value="1"/>
        <param name="robot_name" value="$(arg robot_1)"/>
        <param name="navigation_depth" value="5"/>
      </node>
      <node name="AUV_$(arg robot_2)" pkg="multi_robot_system" type="AUV_robot.py" output="screen" respawn="true" if="$(eval int(arg('number_of_auvs')) >= 3)">
        <param name="robot_ID" value="2"/>
        <param name="robot_name" value="$(arg robot_2)"/>
        <param name="navigation_depth" value="5"/>
      </node>
      <node name="AUV_$(arg robot_3)" pkg="multi_robot_system" type="AUV_robot.py" output="screen" respawn="true" if="$(eval int(arg('number_of_auvs')) >= 4)">
        <param name="robot_ID" value="3"/>
        <param name="robot_name" value="$(arg robot_3)"/>
        <param name="navigation_depth" value="5"/>
      </node>
      <node name="AUV_$(arg robot_4)" pkg="multi_robot_system" type="AUV_robot.py" output="screen" respawn="true" if="$(eval int(arg('number_of_auvs')) >= 5)">
        <param name="robot_ID" value="4"/>
        <param name="robot_name" value="$(arg robot_4)"/>
        <param name="navigation_depth" value="5"/>
      </node>
      <node name="AUV_$(arg robot_5)" pkg="multi_robot_system" type="AUV_robot.py" output="screen" respawn="true" if="$(eval int(arg('number_of_auvs')) >= 6)">
        <param name="robot_ID" value="5"/>
        <param name="robot_name" value="$(arg robot_5)"/>
        <param name="navigation_depth" value="5"/>
      </node>

      <!-- Acoustic communication -->
      <node name="acoustic_communication" pkg="multi_robot_system" type="acoustic_communication.py" output="screen" respawn="true"/>

      <!-- Precompute the ASV node name -->
      <arg name="asv_node_name" default="asv_robot_$(arg number_of_auvs)"/>

      <!-- Precompute the dynamic remap target -->
      <arg name="asv_navigation_remap" default="/robot$(arg number_of_auvs)/navigator/navigation"/>

      <!-- ASV -->
      <node name="$(arg asv_node_name)" pkg="multi_robot_system" type="ASV_robot.py" output="screen" respawn="true" if="$(eval int(arg('number_of_auvs')) >= 2)">
        <param name="robot_ID" value="$(arg number_of_auvs)"/>
        <param name="asv_ID" value="0"/>
        <remap from="/asv0/navigator/navigation" to="$(arg asv_navigation_remap)"/>
      </node>
    </group>
  </group>

  <!-- RViz -->
  <node if="$(arg enableRviz)" pkg="rviz" type="rviz" name="rviz" args="-d $(find multi_robot_system)/scene/MRS.rviz"/>
</launch>
